### PR TITLE
CNV-14623: Hot plugging network interfaces

### PR DESCRIPTION
**Pull request by sjhala-ccs**
Version(s): 4.14+

Issue: [CNV-14623](https://issues.redhat.com//browse/CNV-14623)

Link to docs preview: 

- [Virtctl commands](https://62426--docspreview.netlify.app/openshift-enterprise/latest/virt/getting_started/virt-using-the-cli-tools.html#hot-plug-and-hot-unplug-commands_virt-using-the-cli-tools)
- [Hot plugging secondary network interfaces](https://62426--docspreview.netlify.app/openshift-enterprise/latest/virt/vm_networking/virt-hot-plugging-network-interfaces)

QE review:
- [x] QE has approved this change.

### CODE REVIEW

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+[source,terminal]
+----
+$ virtctl addinterface <vm_name> --network-attachment-definition-name <net_attach_def_name> --name <interface_name>
+----
++
+where:
+
+<vm_name>:: Specifies the name of the `VirtualMachine` object.
+<net_attach_def_name>::  Specifies the name of `NetworkAttachmentDefinition` object.
+<interface_name>:: Specifies the name of the bridge interface.
```
Comment (AlonaKaplan) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:71]: Specifies the VM spec interface/network name.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+. Verify that the new interface is added to the VM specification by checking the virtual machine instance (VMI) status:
```
Comment (AlonaKaplan) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:73]: This step should come after the migration.
After `virtctl addinterface` the interfaces is only added to the VM/VMI specification not to the status.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  {
+    "infoSource": "domain, guest-agent, multus-status",
+    "interfaceName": "eth1",
+    "mac": "02:d8:b8:00:00:2a",
+    "name": "bridge-interface", <1>
+    "queueCount": 1
+  }
+]
+----
+<1> The interface is added to the VMI specification.
```
Comment (AlonaKaplan) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:105]: This is not the VMI specification, it is the status.

File: virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+toc::[]     
+
+You can add or remove secondary network interfaces without stopping your virtual machine (VM). {VirtProductName} supports hot plugging and hot unplugging for Linux bridge interfaces that use the VirtIO device driver.
+
+:FeatureName: Hot plugging and hot unplugging bridge network interfaces
+include::snippets/technology-preview.adoc[]
+
+[id="virtio-limitations_{context}"]
+== VirtIO limitations
+Each VirtIO interface uses one of the limited Peripheral Connect Interface (PCI) slots in the VM. There are a total of 32 slots available. The PCI slots are also used by other devices and must be reserved in advance, therefore slots might not be available on demand. {VirtProductName} reserves up to four slots to be used later for hot plugging interfaces. If your VM has four or more existing interfaces, you cannot hot plug any additional network interfaces.
```
Comment (AlonaKaplan) [inline virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:16]: The suffix " E.g. given two plugged interface, OCP Virt assures 2 more interfaces can be hotplug." is wrong, Openshift virtualization preserves 4, if you have 4 it won't add more. But you may be able to hotplug (q35 preserves 1 interface).

I would rephrase it to - " Four network interfaces are guaranteed to be available for hot pluging, counting in also the existing plugged network interfaces.
 E.g. given two plugged interface, Openshift Virtualization assures 2 more interfaces can be hotplug."

File: virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+:FeatureName: Hot plugging and hot unplugging bridge network interfaces
+include::snippets/technology-preview.adoc[]
+
+[id="virtio-limitations_{context}"]
+== VirtIO limitations
+Each VirtIO interface uses one of the limited Peripheral Connect Interface (PCI) slots in the VM. There are a total of 32 slots available. The PCI slots are also used by other devices and must be reserved in advance, therefore slots might not be available on demand. {VirtProductName} reserves up to four slots to be used later for hot plugging interfaces. If your VM has four or more existing interfaces, you cannot hot plug any additional network interfaces. 
+
+[NOTE]
+====
+The actual number of slots available for hot plugging also depends on the machine type. For example, the default PCI topology for the q35 machine type supports hot plugging a single PCIe device, either emulated or assigned from the host. For more information on PCI topology and hot plug support, see the link:https://libvirt.org/pci-hotplug.html[libvirt documentation].
```
Comment (AlonaKaplan) [inline virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:20]: `the default PCI topology for the q35 machine type supports hot plugging a single PCIe device` - this sentence gives the feeling that with q35 only one interface can be plugged, it is wrong.
I would change to "For example, the default PCI topology for the q35 machine type always adds one extra PCI slot".

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+. Use the following command to hot plug a new network interface to the running VM:
++
+[source,terminal]
+----
+$ virtctl addinterface <vm_name> --network-attachment-definition-name <net_attach_def_name> --name <interface_name>
+----
++
+where:
+
+<vm_name>:: Specifies the name of the `VirtualMachine` object.
```
Comment (apinnick) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:69]: Please use callouts.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        resources:
+          requests:
+            memory: 1024M
+      networks:
+      - name: defaultnetwork
+        pod: {}
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - containerDisk:
+          image: quay.io/kubevirt/fedora-with-test-tooling-container-disk:devel
```
[the remark this answers is not in the document]
Comment (apinnick) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:56]: @sjhala-ccs You could probably use `quay.io/containerdisks/fedora:38`

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc (hunk 6 above)
Comment (sjhala-ccs) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:69]: According to our [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output), using callouts would be impractical in this case because all three variables are on the same line.

Comment (apinnick) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:58]: I do it all the time: "<1> Specify x, y, and z."

Actually, I'm not sure this even needs callouts because the flags are self-evident. 

Suggestion: "$ virtctl addinterface <vm_name> --network-attachment-definition-name <nad_name> --name <interface_name>" with no callouts.

**Comment by AlonaKaplan** [comment]
/lgtm
Thanks!

**Review by yossisegev — CHANGES_REQUESTED** [review]

File: virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+[id="virtio-limitations_{context}"]
+== VirtIO limitations
+Each VirtIO interface uses one of the limited Peripheral Connect Interface (PCI) slots in the VM. There are a total of 32 slots available. The PCI slots are also used by other devices and must be reserved in advance, therefore slots might not be available on demand. {VirtProductName} reserves up to four slots for hot plugging interfaces. This includes any existing plugged network interfaces. For example, if your VM has two existing plugged interfaces, you can hot plug two more network interfaces. 
+
+[NOTE]
+====
+The actual number of slots available for hot plugging also depends on the machine type. For example, the default PCI topology for the q35 machine type supports hot plugging one additional PCIe device. For more information on PCI topology and hot plug support, see the link:https://libvirt.org/pci-hotplug.html[libvirt documentation].
+====
+
+When the VM restarts after you hot plug an interface, that interface becomes part of the standard network interfaces.
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:23]: The is no need to restart. I think the doc says what happen if the user performs a restart, not that it is mandatory to restart.

[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:23]: I have reworded that sentence for clarity.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        devices:
+          interfaces:
+          - masquerade: {}
+            name: defaultnetwork
+       ...
```
Comment (mburke5678) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:42]: s/.../# ...

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+.Prerequisites
+
+* The Multus Container Network Interface (CNI) plugin is deployed on the cluster.
+* A network attachment definition is configured in the same namespace as your VM.
+* You have installed the `virtctl` tool.
+
+.Procedure
+
+. If the VM to which you want to hot plug the network interface is not running, start it by using the following command:
```
Comment (mburke5678) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:19]: This should be a prerequisite.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
++
+[source,terminal]
+----
+$ virtctl start <vm_name>
+----
++
+.Example VM configuration
```
Comment (mburke5678) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:26]: What is this example for?

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+----
++
+where:
+
+<vm_name>:: Specifies the name of the `VirtualMachine` object.
+<net_attach_def_name>::  Specifies the name of `NetworkAttachmentDefinition` object.
+<interface_name>:: Specifies the name of the new network interface.
+
+
+. The `virtctl addinterface` command adds the new network interface to the VM and virtual machine instance (VMI) specification but does not attach it to the running VM. To attach the network interface to the running VM, live migrate the VM by using the following command:
```
Comment (mburke5678) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:63]: I might keep this sentence with the previous step and start this step with _Attach the network interface to_.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+  {
+    "infoSource": "domain, guest-agent, multus-status",
+    "interfaceName": "eth1",
+    "mac": "02:d8:b8:00:00:2a",
+    "name": "bridge-interface", <1>
+    "queueCount": 1
+  }
+]
+----
+<1> The hot plugged interface appears in the VMI status.
```
Comment (mburke5678) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:123]: s/hot plugged/hot-plugged

File: modules/virt-hot-unplugging-bridge-network-interface-cli.adoc [hunk 14]
```diff
@@ -0,0 +1,31 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc              
+
+:_content-type: PROCEDURE                                             
+[id="virt-hot-unplugging-bridge-network-interface_{context}"]                                   
+= Hot unplugging a bridge network interface using the CLI
+
+Remove a bridge network interface from a running virtual machine (VM).
```
Comment (mburke5678) [inline modules/virt-hot-unplugging-bridge-network-interface-cli.adoc:9]: s/Remove/You can remove

File: modules/virt-hot-unplugging-bridge-network-interface-cli.adoc [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+.Procedure
+
+. Hot unplug a bridge network interface by running the following command:
++
+[source,terminal]
+----
+$ virtctl removeinterface <vm_name> --name <interface_name>
+----
+
+. The `virtctl removeinterface` detaches the network interface from the guest, but the interface still exists in the pod. Remove the interface from the pod by migrating the VM:
```
Comment (mburke5678) [inline modules/virt-hot-unplugging-bridge-network-interface-cli.adoc:26]: I might keep the first sentence with the previous step and start this step.

**Comment by mburke5678** [comment]
@sjhala-ccs A few nits, otherwise LGTM. Nice work!

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 16]
```diff
@@ -0,0 +1,123 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc              
+
+:_content-type: PROCEDURE                                             
+[id="virt-hot-plugging-bridge-network-interface_{context}"]                                   
+= Hot plugging a bridge network interface using the CLI
+
+Hot plug a bridge network interface to a virtual machine (VM) while the VM is running.
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:9]: I will update, but I have seen introductory sentences starting with an imperative verb in our docs

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc (hunk 10 above)
Comment (sjhala-ccs) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:19]: I can see that but because this involved the user running a command and a code example, I kept it as a procedure step.

**Comment by sjhala-ccs** [comment]
@yossisegev Can you please approve the PR if everything looks good? Thank you.

File: virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+[id="virtio-limitations_{context}"]
+== VirtIO limitations
+Each VirtIO interface uses one of the limited Peripheral Connect Interface (PCI) slots in the VM. There are a total of 32 slots available. The PCI slots are also used by other devices and must be reserved in advance, therefore slots might not be available on demand. {VirtProductName} reserves up to four slots for hot plugging interfaces. This includes any existing plugged network interfaces. For example, if your VM has two existing plugged interfaces, you can hot plug two more network interfaces. 
+
+[NOTE]
+====
+The actual number of slots available for hot plugging also depends on the machine type. For example, the default PCI topology for the q35 machine type supports hot plugging one additional PCIe device. For more information on PCI topology and hot plug support, see the link:https://libvirt.org/pci-hotplug.html[libvirt documentation].
+====
+
+If you restart the VM after hot plugging an interface, that interface becomes part of the standard network interfaces.
```
Comment (yossisegev) [inline virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:23]: I'm afraid that even after rewording this sentence, it still implies that in order for the hot-plugged interface to become part of the standard network interfaces - you **must** restart the VM.
Perhaps replace the word `becomes` with `remains`?

**Review by yossisegev — CHANGES_REQUESTED** [review]

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+:_content-type: PROCEDURE                                             
+[id="virt-hot-plugging-bridge-network-interface_{context}"]                                   
+= Hot plugging a bridge network interface using the CLI
+
+Hot plug a bridge network interface to a virtual machine (VM) while the VM is running. 
+
+.Prerequisites
+
+* The Multus Container Network Interface (CNI) plugin is deployed on the cluster.
```
Comment (apinnick) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:13]: I just clarified this with Petr Horacek. D/S, no need to mention CNI as prerequisite because it is installed with CNV.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc (hunk 11 above)
Comment (apinnick) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:26]: The code block does not quite fit logically. Unless you state otherwise, the assumption is that the code block is the output of the preceding command. 

Do you really need the code block? It just shows a VM connected to the default pod network. Any VM would look like this when it starts.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+kubevirt-migrate-vm-lj62q   TargetReady       vm-fedora
+kubevirt-migrate-vm-lj62q   Running           vm-fedora
+kubevirt-migrate-vm-lj62q   Succeeded         vm-fedora
+----
+
+. Verify that the new interface is added to the VM by checking the VMI status:
++
+[source,terminal]
+----
+$ oc get vmi vm-fedora -ojsonpath="{ @.status.interfaces }"
```
Comment (apinnick) [inline modules/virt-hot-plugging-bridge-network-interface-cli.adoc:95]: Is `-ojsonpath`  a typo? Looks like it should be `-o jsonpath`.

File: virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+:context: virt-hot-plugging-network-interfaces                        
+                                                                
+toc::[]     
+
+You can add or remove secondary network interfaces without stopping your virtual machine (VM). {VirtProductName} supports hot plugging and hot unplugging for Linux bridge interfaces that use the VirtIO device driver.
+
+:FeatureName: Hot plugging and hot unplugging bridge network interfaces
+include::snippets/technology-preview.adoc[]
+
+[id="virtio-limitations_{context}"]
```
Comment (apinnick) [inline virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:14]: I might be mistaken but I think a peer reviewer recently told me that anchor IDs in assemblies should not contain the context attribute.

File: modules/virt-hot-plugging-bridge-network-interface-cli.adoc (hunk 19 above)
Comment (yossisegev) [reply modules/virt-hot-plugging-bridge-network-interface-cli.adoc:95]: `-ojsonpath` also applies.
Just like `-oyaml`, `-owide`, `oc get pod -n<namespace-name>`, etc.

File: virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc (hunk 17 above)
Comment (AlonaKaplan) [reply virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:23]: But it doesn't remain, it becomes.
For example q35 adds one extra slot. On hotplug this extra slot will be used, but after reboot since the hotplugged interface becomes part of the standard network interfaces, an extra slot will be added.

Comment (yossisegev) [reply virt/virtual_machines/vm_networking/virt-hot-plugging-network-interfaces.adoc:23]: I guess that when 2 people tell you your are drunk - then you should go to sleep (sorry, it sounds better in Hebrew...).
So it's probably me that didn't understand this sentence correctly. Please ignore my comment then, and sorry for the noise.
And thanks for the clarification, Alona.

**Comment by sjhala-ccs** [comment]
@yossisegev Updated the PR. Can you please give an LGTM if everything looks good? Thank you.

**Comment by yossisegev** [comment]
/lgtm
(cnv
